SSO in C3

mansisinha opened the pull request:
# Description
This PR aims to provide support for sso in c3 on cp-ansible

Fixes [ANSIENG-2321](https://confluentinc.atlassian.net/browse/ANSIENG-2321)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

https://jenkins.confluent.io/job/cp-ansible-on-demand/1460/ https://jenkins.confluent.io/job/cp-ansible-on-demand/1459/ https://jenkins.confluent.io/job/cp-ansible-on-demand/1458/

# Checklist:

- [x] Any variable/code changes have been validated to be backwards compatible (doesn't break upgrade)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If required, I have ensured the changes can be discovered by cp-ansible discovery codebase
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules


[ANSIENG-2321]: https://confluentinc.atlassian.net/browse/ANSIENG-2321?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+Groups in JWT
+
+Default:  none
```
Is this the default value of `confluent.oidc.idp.groups.claim.name` setting?

Should we keep it "groups" same as the actual default in code?

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+***
+
+### sso_sub
```
Similar to groups, I feel `claim` / `claim_name` should be added.

mansisinha replied:
do you mean sso_claim_sub?

kc596 replied:
I would suggest `sso_sub_claim`.

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+Sub in JWT
+
+Default:  none
```
The default value of claim is `sub`. Should we use that?

kc596 commented on docs/MOLECULE_SCENARIOS.md:
```diff
@@ -868,6 +868,8 @@ Kafka Broker Customer Listener
 
 RBAC Additional System Admin.
 
+SSO in Control center using Azure
```
OIDC and SAML are 2 well know methods for SSO. And we are using OIDC.
Should we also highlight it here so make it more helpful for future authors / readers?

Something like : "SSO authentication using OIDC in Control center using Azure IdP"

mansisinha replied:
thanks, will do this

kc596 commented on docs/MOLECULE_SCENARIOS.md:
```diff
@@ -878,6 +880,8 @@ Validates that all components on Cluster2 are pointing to the MDS on Cluster1.
 
 Validates that Java 17 is in Use
 
+Validates authenticate api for SSO in Control Center
```
Validates **OIDC** authenticate api for SSO in Control Center ?

kc596 commented on docs/MOLECULE_SCENARIOS.md:
```diff
@@ -1026,6 +1030,8 @@ Kafka Broker Customer Listener.
 
 RBAC Additional System Admin.
 
+SSO in Control center using KeyCloak
```
Similar to suggestion for Azure, could we enhance this to include OIDC?

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+### sso_issuer_url
+
+The issuer url, which is typically the authorization server's URL
```
Should we also add - "This value is used to compare to issuer claim in the JWT token for verification".

If there's some separate doc and we want to keep it short here, it's fine. Just commenting to understand better.

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+### sso_authorize_uri
+
+Base uri for authorize endpoint, that initiates an OAuth authorization request
```
Can remove base uri part.

Can simply be - Endpoint for an OAuth authorization request

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+***
+
+### sso_scope
```
Since this variable is used only to mention scope for groups. And there are many other possible scopes like "openid", "offline_access", etc., it should be renamed to include groups.

mansisinha replied:
would sso_include_groups be good?

kc596 replied:
"scope" is well known term in OAuth. Could we use `sso_groups_scope` ?

kc596 commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+Default:  none
+
+***
+
```
There are other configs as well which have default values but can be configured.

> confluent.oidc.idp.refresh.token.enabled=true

> confluent.oidc.session.token.expiry.ms=900000
> confluent.oidc.session.max.timeout.ms=21600000

We should add variable for above configs as well.

mansisinha replied:
we usually don't add variables for timeouts, does it add any value to make the timeout configurable for the customer? for confluent.oidc.idp.refresh.token.enabled=true it would make sense to add a variable

kc596 replied:
> does it add any value to make the timeout configurable for the customer

I agree - most of customers would not but I think few customer may want to configure that. 
If it can be easily added in future, upon a customer's request, you can skip this for now.

mansisinha replied:
yes, we provide a functinality `kafka_broker_custom_properties` using that, the customer can override any config. The can use it to modify timeout in this case. For refresh token, I think it would add more value to have a variable. Does this sound good?

kc596 replied:
Yes, this seems good.

mansisinha commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,86 @@ Default:  "{{mds_ssl_enabled}}"
+
+***
+
+### sso_groups
```
would changing sso_groups to sso_claim be okay?

kc596 replied:
Could you keep it to `groups_claim` or `sso_groups_claim`?

nsharma-git commented on docs/VARIABLES.md:
```diff
@@ -2380,6 +2380,94 @@ Default:  "{{mds_ssl_enabled}}"
 ***
 
+### sso_mode
+
```
would you like to mention that its not supported for ccs kafka

nsharma-git commented on roles/variables/vars/main.yml:
```diff
@@ -504,6 +504,25 @@ kafka_broker_properties:
       confluent.metadata.server.ssl.truststore.type: BCFKS
       confluent.metadata.server.http2.enabled: false
+  rbac_mds_sso:
+    enabled: "{{ rbac_enabled and not external_mds_enabled and sso_mode != 'none' }}"
```
@kc596 is SSO is configurable only in case of RBAC is enabled ?

kc596 replied:
Yes, RBAC is pre-requisite for SSO because MDS is not enabled if RBAC is not configured.

nsharma-git reviewed (commented):
It would be good if you can add one sample inventory in our examples as this is a new feature in cp-ansible.

mansisinha commented:
> It would be good if you can add one sample inventory in our examples as this is a new feature in cp-ansible.

Added a sample inventory file and incorporated suggested changes. Thanks for the inputs @nsharma-git @kc596

kc596 commented on docs/sample_inventories/rbac_sso_c3.yml:
```diff
@@ -0,0 +1,108 @@
+    sso_authorize_uri: <OAuth authorization endpoint>
+    sso_token_uri: <IdP token endpoint>
+    sso_client_id: <client id>
+    sso_client_password: <client password>
```
Could you also add a corresponding setting for `confluent.oidc.idp.refresh.token.enabled`

mansisinha replied:
added in the sample inventory

wadhwa1 commented on docs/MOLECULE_SCENARIOS.md:
```diff
@@ -1034,6 +1040,8 @@ Validates that MDS is HTTPs on Cluster1 (MDS).
 
 Validates that all components on Cluster2 are pointing to the MDS on Cluster1.
 
+Validates IODC authenticate api for SSO in Control Center
```
typo here - IODC?

wadhwa1 commented on docs/sample_inventories/rbac_sso_c3.yml:
```diff
@@ -0,0 +1,108 @@
+---
+### RBAC SSL - Multiple KSQL Clusters
```
looks like title and description are not updated, we have one more sample inventory `RBAC SSL - Multiple KSQL Clusters`

mansisinha replied:
thanks for pointing out, updated it

wadhwa1 commented on molecule/rbac-mds-kerberos-debian/molecule.yml:
```diff
@@ -261,6 +263,17 @@ provisioner:
+        sso_authorize_uri: https://login.microsoftonline.com/0893715b-959b-4906-a185-2789e1ead045/oauth2/v2.0/authorize
+        sso_token_uri: https://login.microsoftonline.com/0893715b-959b-4906-a185-2789e1ead045/oauth2/v2.0/token
+        sso_client_id: e674334a-de0d-4b28-a9ed-a628285ce7c4
+        sso_client_password: pqb8Q~uNc4AVwLZgXMCE_ONJ6jIEbqnWzp4U_c5c
```
where do we get these client id and password from ?? 
is it ok to keep it hardcoded in the code ?

wadhwa1 commented on molecule/rbac-mds-plain-custom-rhel/molecule.yml:
```diff
@@ -240,6 +242,17 @@ provisioner:
+        sso_jwks_uri: http://ec2-35-166-19-61.us-west-2.compute.amazonaws.com:8080/auth/realms/cpsso/protocol/openid-connect/certs
+        sso_authorize_uri: http://ec2-35-166-19-61.us-west-2.compute.amazonaws.com:8080/auth/realms/cpsso/protocol/openid-connect/auth
+        sso_token_uri: http://ec2-35-166-19-61.us-west-2.compute.amazonaws.com:8080/auth/realms/cpsso/protocol/openid-connect/token
+        sso_client_id: ssologin
```
I see different values for sso clientid and password across different places, can we not re-use them ? 🤔

mansisinha replied:
We have 3 setups for testing sso. Okta, Azure and Keycloak. Each of these has a different id and password

wadhwa1 commented on molecule/rbac-mtls-rhel8/verify.yml:
```diff
@@ -64,3 +65,15 @@
+        validate_certs: false
+        follow_redirects: none
+        status_code: 302
+      register: sso
```
as part of verification, we are testing that we get status code as 302 (which is redirect) - which means SSO is configured correctly and working? do we want to check anything explicitly in registered var - sso `register: sso` ?

also, do we have any negative tests? what happens when we misconfigure SSO details ? 🤔

mansisinha replied:
any result other than 302 would lead to failure of this task. And 302 return only comes when the redirection is successful
